File: OOPS/Class/class_and_object.cpp
/*
Class:

    - Class is a user defined datatype which holds its own 
    members - variables and functions. 
    - The members of a class can be accessed using objects.

Object:

    - Object is an entity with some characteristics and behaviour.
    - Object is an instance of a class and used to access the members of a class.
    - Memory for a class is allocated only when the class is instantiated 
      i.e object for the class is created
*/

#include<iostream>
using namespace std;

// 'class' keyword is used while creating the class
class Person{
    
    //private members can be accessed only within the class
    private:
        string name;
        int age;

    //public members can be accessed from anywhere
    public:
        
        //constructor gets called when memory is allocated for the class
        Person(){
            cout<<"Constructor\n";
        }

        void getDetails(){
            cout<<"Enter name and age ";
            cin>>name>>age;
        }

        void printDetails(){
            cout<<"\nName: "<<name<<" Age: "<<age;
        }
        //destructor gets called when memory is destroyed for the class
        ~Person(){
            cout<<"\nDestructor";
        }
};

int main(){

    //person is the object for the class Person
    Person person;
    
    person.getDetails();
    person.printDetails();

    return 0;
}

/*

Output:
------

Constructor
Enter name and age John 30 

Name: John Age: 30
Destructor

*/

File: OOPS/Class/class_and_object.py
'''
Class:

    - Class is a user defined datatype which holds its own 
    members - variables and functions. 
    - The members of a class can be accessed using objects.

Object:

    - Object is an entity with some characteristics and behaviour.
    - Object is an instance of a class and used to access the members of a class.

'self' keyword:

    - Reference to the current instance of the class
    - Can be user-defined
    - Must be the first argument in the methods
'''

#'class' keyword is used for creating a class
class Person:

    #class attribute
    Temp = ""

    #Initializer is called when a class is instantiated
    def __init__(self):
        print("Initializer")
        
        #instance attributes
        self.name = ""
        self.age = 0

    def getDetails(self):
        self.name = input("Enter name ")
        self.age = int(input("Enter age "))
        
        #accessing class attribute with class name
        Person.Temp = "Hello World"

    def printDetails(self):
        print("Name:",self.name, " Age:",self.age)
        print(Person.Temp)

    #Destructor is called when all the references to the object have been deleted
    def __del__(self):
        print("Destructor")

#person is the object of class Person
person = Person()

person.getDetails()
person.printDetails()


'''
Output:
------

Initializer
Enter name John
Enter age 30
Name: John  Age: 30
Hello World
Destructor

'''